# Тестовое задание
## Парсинг диалогов
Главные задачи, которые должен выполнять скрипт:
- Извлекать реплики с приветствием – где менеджер поздоровался. 
- Извлекать реплики, где менеджер представил себя. 
- Извлекать имя менеджера. 
- Извлекать название компании. 
- Извлекать реплики, где менеджер попрощался.
- Проверять требование к менеджеру: «В каждом диалоге обязательно необходимо поздороваться и попрощаться с клиентом»

In [79]:
import pandas as pd
import re

In [59]:
# данные
data = pd.read_csv('../bewise_test/test_data.csv')
data.sample(5)

dlg_id  line_n     role  \
421       5      84  manager   
166       2       2  manager   
191       2      27  manager   
169       2       5   client   
415       5      78  manager   

                                                  text  
421  А там раз в неделю будет как то диагностироват...  
166                                  Алло здравствуйте  
191                                                Угу  
169                         Ну разговаривать удобно да  
415                       Либо могут отвечать неделями

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 480 entries, 0 to 479
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   dlg_id  480 non-null    int64 
 1   line_n  480 non-null    int64 
 2   role    480 non-null    object
 3   text    480 non-null    object
dtypes: int64(2), object(2)
memory usage: 15.1+ KB


Датафрейм содержит 480 строк и 4 столбца:
- `dlg_id` - номер диалога (0-5)
- `line_n` - номер сообщения (реплики) в диалоге
- `role` - менеджер/клиент
- `text` - текст сообщения 

In [65]:
data.iloc[167]['text']

'Меня зовут ангелина компания диджитал бизнес звоню вам по поводу продления лицензии а мастера мы с вами сотрудничали по видео там'

### 1. Извлечение имени менеджера и названия компании

Для экономии времени извлечь данную информацию из файла можно с использованием уже предобученной модели.

Необходимо выделить следующие типы именованных сущностей:
- `PER` - имя менеджера
- `ORG` - название компании

Каждая реплика состоит из одного предложения (без знаков препинания). При этом, предполагаем, что в реплике максимум может содержаться одно имя и одно название компании.

In [50]:
# !pip install nltk
# !pip install pymorphy2
# nltk.download('punkt')

In [106]:
russian_names = pd.read_csv('../bewise_test/russian_names.csv', sep=';')
russian_names.head()

ID      Name Sex  PeoplesCount     WhenPeoplesCount      Source
0  19903     Аалия   Ж            13  23.06.2016 13:39:41  myData.biz
1  19904  Аанжелла   Ж             0  23.06.2016 13:39:46  myData.biz
2  19905       Аба   Ж          1000  23.06.2016 13:39:55  myData.biz
3  19906      Абав   Ж             0  23.06.2016 13:40:02  myData.biz
4  19907      Абам   Ж            32  23.06.2016 13:40:11  myData.biz

In [107]:
russian_names_list = russian_names['Name'].unique()

In [172]:
def is_self_presentation(message, name):
    if (
        f'меня {name} зовут' in message or
        f'меня зовут {name}' in message or
        f'это {name}' in message
    ):
        return True
    return False

In [109]:
print(is_self_presentation('меня зовут ангелина', 'ангелина'))
text = 'Меня зовут ангелина компания диджитал бизнес звоню вам по поводу продления лицензии а мастера мы с вами сотрудничали по видео там'
print(is_self_presentation(text, 'ангелина'))

True
False


In [179]:
def names_recognize(message):
    morph = pymorphy2.MorphAnalyzer()
    ent_list = []
    
    for word in nltk.word_tokenize(message.lower()):
        is_name = any('Name' in p.tag for p in morph.parse(word))
        if is_name or word == 'китобизнес':
            if word == 'диджитал':
                word = 'диджитал бизнес'
            ent_list.append(word)
    
    manager, company, result = '', '', ''
    for ent in ent_list:
        if is_self_presentation(message.lower(), ent):
            manager = ent
            result += f'manager={manager};'
        elif ent not in russian_names_list and f'компания {ent}' in message:
            company = ent
            result += f'company={company};'
            
    return result   # format: 'manager=... company=...'

In [180]:
# проверка работоспособности функции
print(names_recognize(data.iloc[3]['text']))
print(names_recognize(data.iloc[1]['text']))
print(names_recognize(data.iloc[251]['text']))

manager=ангелина;company=диджитал бизнес;

manager=максим;company=китобизнес;


In [188]:
def is_greeting(message):
    if (
        'добры' in message.lower() or         # добрый день, добрый вечер...
        'здравствуй' in message.lower() or   # здравствуй, здравствуйте
        'привет'in message.lower()           # привет, приветствую
    ):          
        return True

    return False

In [189]:
def is_goodbye_message(message):
    if (
        'свидани' in message.lower() or   # до свидания (и если вдруг ошибка в написании слова)
        'встречи' in message.lower() or   # до встречи 
        'хороше' in message.lower() or    # хорошего дня, хорошего понедельника... НО: не 'хорошо' - это не приветствие
        'добро' in message.lower()        # всего доброго, доброго дня...
    ): 
        return True
    
    return False

In [117]:
def check_manager(greeting, goodbye):
    if greeting and goodbye:
        return True
    return False

In [165]:
def extract_names(name_list, dictionary):
    for elem in name_list:
        if 'manager=' in elem:
            dictionary['manager_name'] = elem.replace('manager=', '').capitalize()
        elif 'company=' in elem:
            dictionary['company_name'] = elem.replace('company=', '').capitalize()
            
    return dictionary

In [169]:
# проверка работоспособности функции
dict_new = extract_names(set(['', 'company=диджитал бизнес', 'manager=ангелина']), {})
print(dict_new['manager_name'])

Ангелина


#### Составим парсер таким образом: 
- **вход**: подаются сообщения диалога в виде таблицы (датафрейма)
- **выход**: печатается
    - имя менеджера
    - имя компании
    - номер и текст реплики приветствия
    - номер и текст реплики само-представления менеджера
    - номер и текст реплики прощания
    - соблюдено ли требование

In [198]:
def parser_dialog(dialog):
    dialog_info = {}
    dtable = dialog.copy()
    dtable['names'] = dtable['text'].apply(names_recognize)
    dtable['greeting'] = dtable['text'].apply(is_greeting)
    dtable['goodbye'] = dtable['text'].apply(is_goodbye_message)
    
    # уникальные имена
    dialog_names = set((';'.join(list(dtable['names'].unique()))).split(';'))
    print(dialog_names)
    
    # извлекаем имя менеджера и компании
    dialog_info = extract_names(dialog_names, dialog_info)
    if 'manager_name' not in dialog_info:
        dialog_info['manager_name'] = 'unknown'
    if 'company_name' not in dialog_info:
        dialog_info['company_name'] = 'unknown'
    
    dtable['self_present'] = dtable.apply(lambda x: is_self_presentation(x['text'].lower(), dialog_info['manager_name'].lower()), axis=1)
    print(dtable.query('self_present')['self_present'])
    
    # извлекаем приветственное сообщение
    dgreeting = dtable.query('greeting')
    dialog_info['greeting_message'] = {row['line_n']: row['text'] for index, row in dgreeting.iterrows()}
    
    # извлекаем прощальное сообщение
    dgoodbye = dtable.query('goodbye')
    dialog_info['goodbye_message'] = {row['line_n']: row['text'] for index, row in dgoodbye.iterrows()}
    
    # извлекаем реплику, где менеджер представляется
    dself = dtable.query('self_present')
    dialog_info['manager_self_presentation'] = {row['line_n']: row['text'] for index, row in dself.iterrows()}
    
    # выполнил ли менеджер поставленное условие (поздороваться и попрощаться)
    dialog_info['requirement_done'] = (dgreeting.shape[0] != 0 and dgoodbye.shape[0] != 0)
    
    return dialog_info

In [199]:
dialogs = {}
for i in data['dlg_id'].unique():
    dialogs[i] = parser_dialog(data.query('dlg_id == @i and role == "manager"'))

{'', 'company=диджитал бизнес', 'manager=ангелина'}
3    True
Name: self_present, dtype: bool
{'', 'company=диджитал бизнес', 'manager=ангелина'}
111    True
Name: self_present, dtype: bool
{'', 'company=диджитал бизнес', 'manager=ангелина'}
167    True
Name: self_present, dtype: bool
{'', 'company=китобизнес', 'manager=максим'}
251    True
Name: self_present, dtype: bool
{''}
Series([], Name: self_present, dtype: bool)
{'', 'manager=анастасия'}
338    True
Name: self_present, dtype: bool


In [200]:
print(dialogs)

{0: {'company_name': 'Диджитал бизнес', 'manager_name': 'Ангелина', 'greeting_message': {1: 'Алло здравствуйте'}, 'goodbye_message': {108: 'Всего хорошего до свидания'}, 'manager_self_presentation': {3: 'Меня зовут ангелина компания диджитал бизнес звоним вам по поводу продления лицензии а мы с серым у вас скоро срок заканчивается'}, 'requirement_done': True}, 1: {'company_name': 'Диджитал бизнес', 'manager_name': 'Ангелина', 'greeting_message': {1: 'Алло здравствуйте'}, 'goodbye_message': {53: 'Угу да вижу я эту почту хорошо тогда исправлю на эту будем ждать ответа всего хорошего', 54: 'До свидания'}, 'manager_self_presentation': {2: 'Меня зовут ангелина компания диджитал бизнес звоню вам по поводу продления а мы сели обратила внимание что у вас срок заканчивается'}, 'requirement_done': True}, 2: {'company_name': 'Диджитал бизнес', 'manager_name': 'Ангелина', 'greeting_message': {2: 'Алло здравствуйте'}, 'goodbye_message': {}, 'manager_self_presentation': {3: 'Меня зовут ангелина комп

In [202]:
def display_inf(inf):
    if type(inf) == dict:
        for key, value in dialog.items():
            print(f'{key}: {display_inf(value)}')
    else:
        print(inf)

In [204]:
for dlg, info in dialogs.items():
    display(info)
    print('-----------')

{'company_name': 'Диджитал бизнес',
 'manager_name': 'Ангелина',
 'greeting_message': {1: 'Алло здравствуйте'},
 'goodbye_message': {108: 'Всего хорошего до свидания'},
 'manager_self_presentation': {3: 'Меня зовут ангелина компания диджитал бизнес звоним вам по поводу продления лицензии а мы с серым у вас скоро срок заканчивается'},
 'requirement_done': True}

-----------


{'company_name': 'Диджитал бизнес',
 'manager_name': 'Ангелина',
 'greeting_message': {1: 'Алло здравствуйте'},
 'goodbye_message': {53: 'Угу да вижу я эту почту хорошо тогда исправлю на эту будем ждать ответа всего хорошего',
  54: 'До свидания'},
 'manager_self_presentation': {2: 'Меня зовут ангелина компания диджитал бизнес звоню вам по поводу продления а мы сели обратила внимание что у вас срок заканчивается'},
 'requirement_done': True}

-----------


{'company_name': 'Диджитал бизнес',
 'manager_name': 'Ангелина',
 'greeting_message': {2: 'Алло здравствуйте'},
 'goodbye_message': {},
 'manager_self_presentation': {3: 'Меня зовут ангелина компания диджитал бизнес звоню вам по поводу продления лицензии а мастера мы с вами сотрудничали по видео там'},
 'requirement_done': False}

-----------


{'company_name': 'Китобизнес',
 'manager_name': 'Максим',
 'greeting_message': {1: 'Алло дмитрий добрый день',
  2: 'Добрый меня максим зовут компания китобизнес удобно говорить'},
 'goodbye_message': {51: 'Угу все хорошо да понедельника тогда всего доброго'},
 'manager_self_presentation': {2: 'Добрый меня максим зовут компания китобизнес удобно говорить'},
 'requirement_done': True}

-----------


{'manager_name': 'unknown',
 'company_name': 'unknown',
 'greeting_message': {},
 'goodbye_message': {33: 'Во вторник все ну с вами да тогда до вторника до свидания'},
 'manager_self_presentation': {},
 'requirement_done': False}

-----------


{'manager_name': 'Анастасия',
 'company_name': 'unknown',
 'greeting_message': {},
 'goodbye_message': {142: 'Ну до свидания хорошего вечера'},
 'manager_self_presentation': {1: 'Да это анастасия'},
 'requirement_done': False}

-----------
